atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/oracle/OracleOverflowQueryFactory.java

```diff
@@ -0,0 +1,88 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.keyvalue.dbkvs.impl.oracle;
+
+import java.util.Collection;
+import java.util.List;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.palantir.atlasdb.keyvalue.dbkvs.impl.FullQuery;
+import com.palantir.atlasdb.keyvalue.dbkvs.impl.OverflowMigrationState;
+import com.palantir.atlasdb.keyvalue.dbkvs.impl.OverflowValue;
+import com.palantir.db.oracle.JdbcHandler;
+import com.palantir.db.oracle.JdbcHandler.ArrayHandler;
+
+public class OracleOverflowQueryFactory extends OracleQueryFactory {
+    private final OverflowMigrationState migrationState;
+
+    public OracleOverflowQueryFactory(String tableName,
+                                      OverflowMigrationState migrationState,
+                                      JdbcHandler jdbcHandler) {
+        super(tableName, jdbcHandler);
+        this.migrationState = migrationState;
+    }
+
+    @Override
+    protected String getValueSubselect(String tableAlias, boolean includeValue) {
+        return includeValue ? ", " + tableAlias + ".val, " + tableAlias + ".overflow " : " ";
+    }
+
+    @Override
+    public boolean hasOverflowValues() {
+        return true;
+    }
+
+    @Override
+    public Collection<FullQuery> getOverflowQueries(Collection<OverflowValue> overflowIds) {
+        List<Object[]> oraRows = Lists.newArrayListWithCapacity(overflowIds.size());
+        for (OverflowValue overflowId : overflowIds) {
+            oraRows.add(new Object[] { null, null, overflowId.id });
+        }
+        ArrayHandler arg = jdbcHandler.createStructArray("PT_MET_CELL_TS", "PT_MET_CELL_TS_TABLE", oraRows);
+        switch (migrationState) {
+        case UNSTARTED:
+            return ImmutableList.of(getOldOverflowQuery(arg));
+        case IN_PROGRESS:
+            return ImmutableList.of(getOldOverflowQuery(arg), getNewOverflowQuery(arg));
+        case FINISHING: // fall through
+        case FINISHED:
+            return ImmutableList.of(getNewOverflowQuery(arg));
+        default:
+            throw new EnumConstantNotPresentException(OverflowMigrationState.class, migrationState.name());
+        }
+    }
+
+    private FullQuery getOldOverflowQuery(ArrayHandler arg) {
+        String query =
+                " /*SQL_MET_SELECT_OVERFLOW */ " +
+                " SELECT /*+ USE_NL(t o) LEADING(t o) INDEX(o pk_pt_metropolis_overflow) */ " +
+                "   o.id, o.val " +
+                " FROM pt_metropolis_overflow o, TABLE(CAST(? AS PT_MET_CELL_TS_TABLE)) t " +
+                " WHERE t.max_ts = o.id ";
+        return new FullQuery(query).withArg(arg);
+    }
+
+    private FullQuery getNewOverflowQuery(ArrayHandler arg) {
```

Comment:
would be massively preferable if all new customers on oss atlas did not have overflow tables, they are a huge headache (and, similar reasoning behind them not having upgrade tasks, plus an upgrade task to re-integrate random large values from all over the tables is going to be very long-running and annoying)